File: services/engine/engine/schema/testable.py
"""The shared testable-validator, Python twin of packages/schema/src/testable.ts.

Rules come from packages/schema/testable.rules.json; this only interprets them."""

import json
from functools import lru_cache
from typing import Any

from engine import config


@lru_cache(maxsize=1)
def _rules() -> list[dict]:
    return json.loads((config.SCHEMA_DIR / "testable.rules.json").read_text())["rules"]


def _resolve(doc: Any, pointer: str) -> Any:
    cur = doc
    for seg in pointer.split("/")[1:]:
        if not isinstance(cur, dict):
            return None
        cur = cur.get(seg)
    return cur


def _present(v: Any, not_: Any = None) -> bool:
    if v is None or v == "" or (isinstance(v, list) and not v):
        return False
    return not (not_ is not None and v == not_)


def check_testable(spec: dict) -> dict:
    """Returns {"testable": bool, "missing": [sentence], "missing_ids": [rule id]}."""
    missing, ids = [], []
    for rule in _rules():
        values = [_resolve(spec, p["path"]) for p in rule["paths"]]
        if rule["kind"] == "present_any":
            ok = any(_present(v, p.get("not")) for v, p in zip(values, rule["paths"], strict=True))
        else:
            ok = not _present(values[0])
        if not ok:
            vals = "; ".join(map(str, values[0])) if isinstance(values[0], list) else str(values[0] or "")
            missing.append(rule["missing"].replace("{values}", vals))
            ids.append(rule["id"])
    return {"testable": not missing, "missing": missing, "missing_ids": ids}

File: packages/schema/src/testable.ts
// The shared testable-validator. The rules live in ../testable.rules.json; this is a ~25-line interpreter.
// engine/schema/testable.py is the Python twin. Both run fixtures/testable.cases.json.
import rules from "../testable.rules.json";

type PathSpec = { path: string; not?: unknown };
type Rule = { id: string; kind: "present_any" | "empty"; paths: PathSpec[]; missing: string };

export interface TestableResult { testable: boolean; missing: string[]; missing_ids: string[] }

function resolve(doc: unknown, pointer: string): unknown {
  let cur: unknown = doc;
  for (const seg of pointer.split("/").slice(1)) {
    if (cur === null || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[seg];
  }
  return cur;
}

function present(v: unknown, not?: unknown): boolean {
  if (v === null || v === undefined || v === "") return false;
  if (Array.isArray(v) && v.length === 0) return false;
  if (not !== undefined && v === not) return false;
  return true;
}

export function checkTestable(spec: unknown): TestableResult {
  const missing: string[] = [];
  const missing_ids: string[] = [];
  for (const rule of (rules as { rules: Rule[] }).rules) {
    const values = rule.paths.map((p) => resolve(spec, p.path));
    const ok = rule.kind === "present_any"
      ? rule.paths.some((p, i) => present(values[i], p.not))
      : !present(values[0]);
    if (!ok) {
      const vals = Array.isArray(values[0]) ? (values[0] as unknown[]).join("; ") : String(values[0] ?? "");
      missing.push(rule.missing.replace("{values}", vals));
      missing_ids.push(rule.id);
    }
  }
  return { testable: missing.length === 0, missing, missing_ids };
}
